src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -442,9 +443,9 @@ internal static void OnValuesChanged(ExecutionContext previousExecutionCtx, Exec
                 else // Implied: nextChangeNotifications != null
                 {
                     // Notifications can't exist without values
-                    Debug.Assert(nextExecutionCtx.m_localValues != null);
+                    Debug.Assert(nextExecutionCtx!.m_localValues != null); // TODO-NULLABLE: Compiler can't see that we're only here when this is non-null
                     // No previous values, so just check current against null
-                    foreach (IAsyncLocal local in nextChangeNotifications)
+                    foreach (IAsyncLocal local in nextChangeNotifications!) // TODO-NULLABLE: Compiler can't see that we're only here when this is non-null
```

Comment:
I'm surprised about this one -- I would expect the compiler to assume from the `&&` and `else if` above to understand this would be not nullable.

cc: @jaredpar @gafter @cston

Reply:
Opened https://github.com/dotnet/roslyn/issues/34665